Lib/fontTools/ttLib/tables/TupleVariation.py

```diff
@@ -446,6 +444,125 @@ def getTupleSize_(flags, axisCount):
+					for d in self.coordinates
+				]
+		else:
+			if scalar == 0:
+				self.coordinates = [0] * len(self.coordinates)
+			else:
+				self.coordinates = [
+					d * scalar if d is not None else None
+					for d in self.coordinates
+				]
+
+	def roundDeltas(self):
+		deltaType = self.getDeltaType()
+		if deltaType == "gvar":
+			self.coordinates = [
+				(otRound(d[0]), otRound(d[1])) if d is not None else None
+				for d in self.coordinates
+			]
+		else:
+			self.coordinates = [
+				otRound(d) if d is not None else None for d in self.coordinates
+			]
+
+	def calcInferredDeltas(self, origCoords, endPts):
+		from fontTools.varLib.iup import iup_delta
+
+		if self.getDeltaType() == "cvar":
+			raise TypeError(
+				"Only 'gvar' TupleVariation can have inferred deltas"
+			)
+		if None in self.coordinates:
+			if len(self.coordinates) != len(origCoords):
+				raise ValueError(
+					"Expected len(origCoords) == %d; found %d"
+					% (len(self.coordinates), len(origCoords))
+				)
+			self.coordinates = iup_delta(self.coordinates, origCoords, endPts)
+
+	def optimize(self, origCoords, endPts, tolerance=0.5, isComposite=False):
+		from fontTools.varLib.iup import iup_delta_optimize
+
+		if None in self.coordinates:
+			return  # already optimized
+
+		deltaOpt = iup_delta_optimize(
+		    self.coordinates, origCoords, endPts, tolerance=tolerance
+		)
+		if None in deltaOpt:
+			if isComposite and all(d is None for d in deltaOpt):
+				# Fix for macOS composites
+				# https://github.com/fonttools/fonttools/issues/1381
+				deltaOpt = [(0, 0)] + [None] * (len(deltaOpt) - 1)
+			# Use "optimized" version only if smaller...
+			varOpt = TupleVariation(self.axes, deltaOpt)
+
+			# Shouldn't matter that this is different from fvar...?
+			axisTags = sorted(self.axes.keys())
+			tupleData, auxData, _ = self.compile(axisTags, [], None)
+			unoptimizedLength = len(tupleData) + len(auxData)
+			tupleData, auxData, _ = varOpt.compile(axisTags, [], None)
+			optimizedLength = len(tupleData) + len(auxData)
+
+			if optimizedLength < unoptimizedLength:
+				self.coordinates = varOpt.coordinates
+
+	def __iadd__(self, other):
+		if not isinstance(other, TupleVariation):
+			return NotImplemented
+		deltas1 = self.coordinates
+		length = len(deltas1)
+		deltaType = self.getDeltaType()
+		deltas2 = other.coordinates
+		if len(deltas2) != length:
+			raise ValueError(
+				"cannot sum TupleVariation deltas with different lengths"
+			)
+		if deltaType == "gvar":
+			for i, d2 in zip(range(length), deltas2):
+				d1 = deltas1[i]
+				try:
+					deltas1[i] = (d1[0] + d2[0], d1[1] + d2[1])
+				except TypeError:
+					raise ValueError(
+						"cannot sum gvar deltas with inferred points"
+					)
+		else:
+			for i, d2 in zip(range(length), deltas2):
+				d1 = deltas1[i]
+				if d1 is not None and d2 is not None:
+					deltas1[i] = d1 + d2
+				elif d1 is None and d2 is not None:
+					deltas1[i] = d2
+				# elif d2 is None do nothing
```

Comment:
Here again we're writing code twice to handle one vs two values; lets figure out a way to not do that. For example, we could have a width-aware add fn. I vaguely imagine you end up with something like:

```
if their deltas length != ours err
if thier coord width != ours err [I think we're missing this currently?]
deltas = [_addCoord(d1, d2) for d1, d2 in zip(self.coordinates, other.coordinates)]
# _addCoord can get and use the width and do the right thing
```

Reply:
> if thier coord width != ours err [I think we're missing this currently?]

no, it is possible that `self.getCoordWidth() == 0` but `other.getCoordWidth() != 0`, e.g. all self.coordinates are None but other's aren't, in which case adding self with the other gives the other.

Reply:
Another reason for all the branching is that the `__iadd__` method needs to deal with the different meanings that the `None` coordinates values have in gvar vs cvar TupleVariations: within the gvar, when deltas are not provided explicitly for some points, they need to be inferred; however, this does not apply for the 'cvar' table, where if deltas are not provided for some CVT values, then no adjustments are made (i.e. `None` is same as 0).
So when we have a gvar-style TupleVariation (coordWidth=2) we cannot meaningful "add" its coordinates to another such TupleVariation unless we have first computed all the inferred None values (hence the defensive ValueError when that's not the case).
When the `None` values occur in a cvar-style TupleVariation (coordWidth=1) we can "add" them up by effectively treating `None` as a 0.

Reply:
and the reason I don't bother checking if we are adding TupleVariation of different width is that it is unlikely that one ever tries to sum a TupleVariation from gvar with another from cvar or viceversa.
These two flavors of TupleVariation really should be different types, rather than same class parametrised by the type of its content (tuple of numbers vs plain numbers). But I didn't write this code originally and I wanted to make the new instancer's code more or less self-contained without introducing too sweeping changes to the rest of the library.

Reply:
ok, I can add a comment.